### PR TITLE
Grid insert instead of swap

### LINKED ISSUE #266: Grid layout drag and drop
Currently the grid layout swaps tabs on drag and drop. This means that moving a tab group from the bottom of the grid to the top without messing up the order of the others requires as many drag and drops as there are tab groups. In my case - that's over 50.

Grid layout drag and drop should insert at the dropped position, not swap. This way swapping takes 2 drag and drops, and inserting only takes 1 (As opposed to 50)

### PR DESCRIPTION
**Pull request by jnvsor**
Fixes #266 

Your complaint was that inserting instead of swapping was too slow due to having to reflow all the tabs between them:

> I tried that behavior at first, and it made things so jumpy it was just incredibly confusing to follow.

> it becomes increasingly taxing on the browser to animate/move properly, becoming also increasingly harder to follow

> Moving so many elements (all groups AND all tabs) is very taxing on the browser. Things lag, jump around...

You only need to rearrange a group's tabs if it's changing row - otherwise all elements will have the same dimensions. This means a maximum of 4 calls to arrange tabs per drag 'n drop. (As opposed to 2 for swap method)

For reference, with 285 tabs divided by 15 groups (The only one with less than 20 made that size to trigger a rearrange) finishes reflowing in (Eyeballing it here) less than half a second. No jumping, no confusion. Just works.

### CODE REVIEW

**Comment by Quicksaver** [comment]
I can't accept this behavior. My other concern was that highlighting the "target" group implies that the moving group will land **at that place** and not before or after as you suggest. This doesn't address that at all.

**Comment by jnvsor** [comment]
> My other concern was that highlighting the "target" group implies that the moving group will land at that place and not before or after as you suggest

It does. Though it might be hard to see the group names on that low Q gif

**Comment by Quicksaver** [comment]
Hmm, I don't think you understood me. When you drag a group over another group, it makes it darker. This implies that something will happen between those groups that have the darker shade, but there is absolutely no indication that the group being dragged will be placed before or after the group under the cursor, so the logical assumption is that the groups will be switched.

I don't see how the group names have anything to do with it...

**Comment by jnvsor** [comment]
It doesn't imply a swap IMO (Perhaps a poll or some UX guys are in order)

> but there is absolutely no indication that the group being dragged will be placed before or after the group under the cursor

Just to be clear - the group you are dragging ends up *under the cursor, in the exact position of the highlighted group.* The highlighted group and groups between then move along into the new vacancy from the original drag.

Could you try the actual behaviour to see how it feels?

**Comment by Quicksaver** [comment]
> the group you are dragging ends up under the cursor, in the exact position of the highlighted group

I can't argue this point, it does make sense. Please do fix it if you think there's an off-by-one error currently, and I'll be sure to try it out when I can. :)

**Comment by iShyboy** [comment]
We have the elements of free arrange, you can use it on grid layout maybe?
look:
c
Don't mind the version, just I don't have the time to switch FF now.
Look at the line, the group shouldn't be highlighted when inserting, show this traced line instead.
Highlight the group only when swapping, the margins for this isn't my job. :)

**Comment by jnvsor** [comment]
There's definitely an off by one error, but it may be in something else entirely. It seems to appear when dragging a tab out of a group that only has one tab in it. This group is then automatically removed but the slots aren't reordered (And my behaviour assumes they would be)

Is this a bug in preexisting code or should I not assume tab slots are sequential?

**Comment by Quicksaver** [comment]
@iShyboy No. Just, don't do that. :) Those trenches (that's the name the code uses for them anyway) are really more a nuisance than anything, and I only didn't already get rid of them because they're still useful in free arrange. The less they're used, the better for later. ;)

**Comment by Quicksaver** [comment]
> Is this a bug in preexisting code or should I not assume tab slots are sequential?

(I hadn't checked your code yet.) Indeed, don't assume that, just iterate all groups and compare/update their slots accordingly. It's very easy to end up with gaps in the slot numbers.

**Comment by iShyboy** [comment]
Nevermind, is just that those lines remember me the grid lines of home from android launchers, they appear when you want to insert something, but the highlight is when you want to make a folder. It's the same.

**Comment by jnvsor** [comment]
> Indeed, don't assume that, just iterate all groups and compare/update their slots accordingly. It's very easy to end up with gaps in the slot numbers.

Could I cheat with `reorderTabItemsBasedOnTabOrder` or something similar? If not I'll have to rewrite the commit entirely and it's half past midnight here so that would be tomorrow

**Comment by Quicksaver** [comment]
I won't be able to even check things out for at least two weeks, so you have plenty of time to perfect it. ;)

**Comment by jnvsor** [comment]
Allrighty, this seems to work without any off-by-one errors. Added some comments too since I was getting confused myself there for a bit.

Would anyone like to test this? [Here's a gifycat](https://gfycat.com/PointlessGraveHarborporpoise) demonstrating the working (As far as I can tell) behaviour with nice big monocolor tabs

File: resource/modules/Drag.jsm [hunk 1]
```diff
@@ -421,22 +421,64 @@ this.GroupDrag.prototype = {
...
-		this.item.slot = targetSlot;
-		this.item.row = targetRow;
-		this.item._gridBounds = targetBounds;
-		this.dropTarget.save();
-		this.dropTarget.arrange();
-		this.item.save();
-		this.item.arrange();
+		// Move the dragged group to the slot and shift everything in between
+		// There's no need to recalc the grid dimensions, they should
+		// stay the same, only the groups that change row change size
```
Comment (Quicksaver) [inline resource/modules/Drag.jsm]: Can you remove this comment? It's no longer true as all groups inbetween move now.

Comment (jnvsor) [inline resource/modules/Drag.jsm]: > Move the dragged group to the slot and shift everything in between

Already did :confused:

Comment (Quicksaver) [inline resource/modules/Drag.jsm]: Then you must have not updated the commit (have you?) because I still see it.

Comment (jnvsor) [inline resource/modules/Drag.jsm]: That's very strange. Here's the comment I'm seeing:

> Move the dragged group to the slot and shift everything in between
> There's no need to recalc the grid dimensions, they should
> stay the same, only the groups that change row change size

commit ebc7b2a0c6d

**Comment by Quicksaver** [comment]
Disregard the comments about the comment (I already deleted them). I'm a huge idiot.

**Comment by jnvsor** [comment]
Aren't we all? :)

**Comment by Quicksaver** [comment]
Some of us have unusual high peaks of the trait at times. :)

BTW I won't get a chance to actually look into this for at least a couple of weeks. So don't be discouraged if this just sits here for a while.

**Comment by B00ze64** [comment]
I'm also interested in this, it makes it behave like other things in the O/S, like when you drag a bookmark or a menu item up/down, it gets inserted there rather than swap places. You could draw a black bar in between groups when you drag, rather than shade the group under the cursor, if you wanted to make it behave even more like other programs.

PS: Luis I do NOT know how you have ANY time to work on this (and you've worked on it so much) - Today I have ~90 emails from GitHub regarding this addOn alone; where is the time to do anything else?! You must have a master's degree in time management ;-)

**Comment by Quicksaver** [comment]
> PS: Luis I do NOT know how you have ANY time to work on this (and you've worked on it so much) - Today I have ~90 emails from GitHub regarding this addOn alone; where is the time to do anything else?! You must have a master's degree in time management ;-)

Sometimes I think I do too. ;) And luckily every once in a while, someone awesome like @jnvsor will help out like in this case.

**Comment by jnvsor** [comment]
Had a chance to take a look at this yet @Quicksaver ?

**Comment by Quicksaver** [comment]
Not yet, you'll know when I do.

**Comment by Spitfire1900** [comment]
👍

**Comment by jnvsor** [comment]
Done!

**Comment by jnvsor** [comment]
Sweet! When can I expect it to show up in my browsers? :)

**Comment by Quicksaver** [comment]
It might take a few weeks, there's a crap ton of bugs to fix. It'll take a while to go through everything properly. 😜

**Pull request merged**